src/kbc_modules/offline_fs_kbc/mod.rs

```diff
@@ -0,0 +1,204 @@
+            kbs_info: HashMap::new(),
+            keys: load_keys(KEYS_PATH).map_err(|e| anyhow!("Failed to load keys: {}", e)),
+            ciphers: ciphers(),
+        }
+    }
+}
+
+fn load_keys(keyfile_name: &str) -> Result<Keys> {
+    let keys_json = fs::read_to_string(keyfile_name)?;
+    // Redact parsing errors to avoid side-channels
+    let encoded_keys: HashMap<String, String> =
+        serde_json::from_str(&keys_json).map_err(|_| anyhow!("Failed to parse keys JSON file"))?;
+    encoded_keys
+        .iter()
+        .map(|(k, v)| match decode(v) {
+            Ok(key) => Ok((k.clone(), key)),
+            Err(_) => Err(anyhow!("Failed to decode key")),
+        })
+        .collect()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use base64::encode;
+    use openssl::symm::encrypt;
+    use std::env;
+    use std::path::Path;
+
+    const KID: &str = "foo";
+    const KEY: [u8; 32] = *b"passphrasewhichneedstobe32bytes!";
+
+    #[test]
+    fn test_decrypt_payload() {
+        let iv = b"ivmustbe16bytes!";
+        let data = b"bar";
+
+        let cipher_key = "aes_256_ctr";
+        let cipher = ciphers().get(cipher_key).unwrap().to_owned();
+        let wrapped_data = encrypt(cipher, &KEY, Some(iv), data).unwrap();
+
+        let annotation = format!(
+            "{{
+    \"kid\": \"{}\",
+    \"wrapped_data\": {:?},
+    \"iv\": {:?},
+    \"wrap_type\": \"{}\"
+}}",
+            KID, wrapped_data, iv, cipher_key
+        );
+
+        let mut kbc = OfflineFsKbc {
+            kbs_info: HashMap::new(),
+            keys: Ok([(KID.to_string(), KEY.to_vec())].iter().cloned().collect()),
+            ciphers: ciphers(),
+        };
+
+        assert_eq!(kbc.decrypt_payload(&annotation).unwrap(), data);
+
+        let invalid_annotation = &annotation[..annotation.len() - 1];
+        assert!(kbc.decrypt_payload(invalid_annotation).is_err());
+
+        let mut key_load_failure_kbc = OfflineFsKbc {
+            kbs_info: HashMap::new(),
+            keys: Err(anyhow!("")),
+            ciphers: ciphers(),
+        };
+        assert!(key_load_failure_kbc.decrypt_payload(&annotation).is_err());
+
+        let mut unknown_kid_kbc = OfflineFsKbc {
+            kbs_info: HashMap::new(),
+            keys: Ok([(String::from("baz"), KEY.to_vec())]
+                .iter()
+                .cloned()
+                .collect()),
+            ciphers: ciphers(),
+        };
+        assert!(unknown_kid_kbc.decrypt_payload(&annotation).is_err());
+
+        // Notice that a wrong key does not yield an error
+        let mut invalid_key_kbc = OfflineFsKbc {
+            kbs_info: HashMap::new(),
+            keys: Ok([(
+                KID.to_string(),
+                b"thispassphraseisntactually32bytes".to_vec(),
+            )]
+            .iter()
+            .cloned()
+            .collect()),
+            ciphers: ciphers(),
+        };
+        assert!(invalid_key_kbc.decrypt_payload(&annotation).is_err());
```

Comment:
The comment in line 164 says "does not yield an error" but the assert at the bottom checks that .is_err() . Which is correct?

Reply:
Sorry, I still miss something. Why does decrypt_payload() here (line 176) return Err and not Ok(bad_decrypted_data) ?